### ISSUE TITLE
Shutdown at specific epoch when applying new consensus rules

### ISSUE THREAD

**Issue by justinmoon** [fedimint/fedimint#948]
When making an upgrade, we need all peers to shutdown at the same epoch. Otherwise, different peers might end up in different states.

Then everyone switches out fedimintd and applies the new rules at the same time.

**Comment by jkitman** [comment]
Peers can signal a desire to shutdown, perhaps passing in an optional consensus config hash, and when a threshold of peers are signaling in consensus, we automatically shutdown.

**Comment by dpc** [comment]
Shutdown and then get restarted automatically by systemd? I don't think the shutdown changes anything here. If anything the fedimintd should just refuse to generate any new epochs and wait (for operator, or some other things unblocking it).

I guess we either invest into writing proper "migration" logic, where fedimintd can support both the previous and new consensus rules, and switch at a given (probably `ConsensusItem` driven) epoch automatically (more work for fedimint devs and module developers, smoother experience for the guardians) or just stop and have all the guardians coordinate simultanous restart.

I wonder if we could even have an easy hybrid. We could have two sets of consensus ports. Before upgrade the guardian would run the fedimintd like this:


```
fedimintd-0.14 --upgrade-new-bin fedimintd-0.15
```

After reaching upgrade epoch, fedimintd-0.14 would restart itself. On start it would detect that it is already at upgrade epoch and start operating in a read-only mode on an alternative set of ports, it would also start `fedimintd-0.15` that would take over normal operation.

Each fedimintd would know that if it is connecting to a peer that reports higher versions then their own, they should try alternative set of ports.

This way the old version can stick around indefinitely and allow any peers that were down etc. to finish syncing to the upgrade epoch when possible. While at no point we need a binary that can support both old and new protocol.

**Comment by jkitman** [comment]
> After reaching upgrade epoch, fedimintd-0.14 would restart itself. On start it would detect that it is already at upgrade epoch and start operating in a read-only mode on an alternative set of ports, it would also start fedimintd-0.15 that would take over normal operation.

Yeah, I was imagining something similar like that would be good to minimize downtime, since the guardians might not be able to manually coordinate the upgrade within a short time window.

**Comment by elsirion** [comment]
I don't think we should focus on uncoordinated upgrades too much yet. Being able to upgrade *at all* would be a huge improvement and is far lower complexity. Adding automatic switchover or even running two federations in parallel can build on top of that and would require far more work right now imo.

**Comment by jkitman** [comment]
@elsirion What do you think of this for the MVP upgrade:

1. Guardians modify the consensus config to change version number or other consensus values
2. Hash of config is sent as a consensus item `ConsensusUpgrade(hash)`
3. Once a threshold of peers is signaling for the same upgrade, automatically stop consensus at that epoch and shutdown
4. If version changed or other items such as pubkeys, start the new binary and do any required DB migration
5. Guardians connect to each other with the new consensus config and validate with peers it is correct

For version upgrades, we could eventually automate the steps with an upgrade script.

**Comment by elsirion** [comment]
@jkitman First modifying config and then stopping consensus at a certain epoch doesn't work imo: you need the old cfg to keep running till that epoch. We don't need to automate this for now, we can just have an admin command that tells the fedimintd to shut down after epoch x:

1. Every guardian tells their `fedimintd` to stop at epoch x (e.g. during epoch ~x-100 so that there's enough time)
2. All `fedimintd`s stop at epoch x
3. Everyone upgrades the config and possibly fedimint binaries
4. `fedimintd` is restarted, applies DB migrations and continues to run from epoch x+1 with new consensus rules

There will be downtime, but there's so much less that can go wrong. Specifying and testing a water tight version of your on-the-fly upgrade approach will be much more involved imo and could introduce quite some instability before we even have an MVP.

**Comment by jkitman** [comment]
@elsirion I meant modify a copy of the consensus config, rather than modify the existing config in-place.  A script could replace the existing config with the modified one once enough guardians agree.

**Comment by jkitman** [comment]
@elsirion Was working on this and realized we don't really have an authenticated way for a guardian to send messages to their `fedimintd` server, do we?

**Comment by dpc** [comment]
> @elsirion Was working on this and realized we don't really have an authenticated way for a guardian to send messages to their `fedimintd` server, do we?

My previous attempt at that #1171 . I never had time to get back to it, but there's a discussion about the design there. @jkitman

**Comment by elsirion** [comment]
Ah, so we did post something on GH! I just didn't search in PRs …

**Comment by justinmoon** [comment]
dev call: we might close this issue as "not planned"

**Comment by elsirion** [comment]
We have the functionality now, but since 0.5 we don't even want to use that anymore since coordinating a shutdown/upgrade/restart is a DevOps nightmare.
